server/storage/schema/wal.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/pkg/v3/pbutil"
+	"go.etcd.io/etcd/raft/v3/raftpb"
+	"go.etcd.io/etcd/server/v3/wal"
+)
+
+// EtcdVersionFromWAL returns minimal etcd storage able to interpred provided wal log,
+// determined by looking on entries since last snaphost and returning highest
+// etcd version annotation from used messages, fields, enums and their values.
+func EtcdVersionFromWAL(wal *wal.WAL) *semver.Version {
```

Comment:
Nit suggestion: Naming of this function makes me think we are only fetching the version, but we also calculate the max version, so maybe the name should be `MinimalStorageVersionFromWAL` or something like that.

Reply:
Nice name! I like it